credentials/generate-revocation-set.py

```diff
@@ -0,0 +1,243 @@
+    """DCL PAA mirroring tools"""
+
+    production = False
+    dcld = use_test_net_dcld
+
+    if len(use_main_net_dcld) > 0:
+        dcld = use_main_net_dcld
+        production = True
+
+    use_rest = use_main_net_http or use_test_net_http
+    if use_main_net_http:
+        production = True
+
+    rest_node_url = PRODUCTION_NODE_URL_REST if production else TEST_NODE_URL_REST
+
+    if use_rest:
+        revocation_point_list = requests.get(f"{rest_node_url}/dcl/pki/revocation-points").json()["PkiRevocationDistributionPoint"]
+    else:
+        cmdlist = ['config', 'output', 'json']
+        subprocess.Popen([dcld] + cmdlist)
+
+        cmdlist = ['query', 'pki', 'all-revocation-points']
+
+        cmdpipe = subprocess.Popen(use_dcld(dcld, production, cmdlist), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        revocation_point_list = json.loads(cmdpipe.stdout.read())["PkiRevocationDistributionPoint"]
+
+    revocation_set = []
+
+    for revocation_point in revocation_point_list:
+        # 1. Validate Revocation Type
+        if revocation_point["revocationType"] != RevocationType.CRL:
+            continue
+
+        # 2. Parse the certificate
+        crl_signer_certificate = x509.load_pem_x509_certificate(revocation_point["crlSignerCertificate"])
+
+        vid = revocation_point["vid"]
+        pid = revocation_point["pid"]
+        is_paa = revocation_point["isPAA"]
+
+        # 3. && 4. Validate VID/PID
+        crl_vid = extract_single_integer_attribute(crl_signer_certificate.subject, OID_VENDOR_ID)
+        crl_pid = extract_single_integer_attribute(crl_signer_certificate.subject, OID_PRODUCT_ID)
+
+        if is_paa == True:
+            if crl_vid != None:
+                if vid != crl_vid:
+                    continue
+        else:
+            if vid == None or vid != crl_vid:
+                continue
+            if crl_pid != None:
+                if pid != crl_pid:
+                    continue
+
+        # 5.
+        crl_signer_issuer_name = base64.b64encode(crl_signer_certificate.issuer.public_bytes()).decode('utf-8')
+
+        crl_signer_authority_key_id = crl_signer_certificate.extensions.get_extension_for_oid(
+            x509.OID_AUTHORITY_KEY_IDENTIFIER).value.key_identifier
+
+        paa_certificate = None
+
+        if use_rest:
+            response = requests.get(
+                f"{rest_node_url}/dcl/pki/certificates/{crl_signer_issuer_name}/{crl_signer_authority_key_id}").json()["approvedCertificates"]["certs"][0]
+            paa_certificate = response["pemCert"]
+        else:
+            cmdlist = ['query', 'pki', 'x509-cert', '-u', crl_signer_issuer_name, '-k', crl_signer_authority_key_id]
+            cmdpipe = subprocess.Popen(use_dcld(dcld, production, cmdlist), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            paa_certificate = json.loads(cmdpipe.stdout.read())["approvedCertificates"]["certs"][0]["pemCert"]
```

Comment:
Extract this operation to a helper function.  Can be done as a follow-up.

Reply:
Added comments in the src file to capture the follow-up work.